// rclnodejs/generated/action_tutorials_interfaces/action_tutorials_interfaces__action__Fibonacci.js
'use strict';
class action_tutorials_interfaces__action__Fibonacci {
  static get Goal() {
    return require('./action_tutorials_interfaces__action__Fibonacci_Goal.js');
  }
  static get GoalRes() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get Cancel() {
    return require('../action_msgs/action_msgs__srv__CancelGoalRequest.js');
  }
  static get CancelRes() {
    return require('../action_msgs/action_msgs__srv__CancelGoalResponse.js');
  }
  static get ResultReq() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get ResultRes() {
    return require('./action_tutorials_interfaces__action__Fibonacci_Result.js');
  }
  static get Feedback() {
    return require('./action_tutorials_interfaces__action__Fibonacci_Feedback.js');
  }
  static get Status() {
    return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
  }
  static type() {
    return {pkgName: 'action_tutorials_interfaces', subFolder: 'action', interfaceName: 'Fibonacci'};
  }
}
module.exports = action_tutorials_interfaces__action__Fibonacci;
