// rclnodejs/generated/turtlesim/turtlesim__action__RotateAbsolute.js
'use strict';
class turtlesim__action__RotateAbsolute {
  static get Goal() {
    return require('./turtlesim__action__RotateAbsolute_Goal.js');
  }
  static get GoalRes() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get Cancel() {
    return require('../action_msgs/action_msgs__srv__CancelGoalRequest.js');
  }
  static get CancelRes() {
    return require('../action_msgs/action_msgs__srv__CancelGoalResponse.js');
  }
  static get ResultReq() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get ResultRes() {
    return require('./turtlesim__action__RotateAbsolute_Result.js');
  }
  static get Feedback() {
    return require('./turtlesim__action__RotateAbsolute_Feedback.js');
  }
  static get Status() {
    return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'action', interfaceName: 'RotateAbsolute'};
  }
}
module.exports = turtlesim__action__RotateAbsolute;
