// rclnodejs/generated/nav2_msgs/nav2_msgs__action__FollowWaypoints_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const StructType = require('ref-struct');
const ref = require('ref');
const ArrayType = require('ref-array');
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GoalStatusWrapper = require('../../generated/action_msgs/action_msgs__msg__GoalStatus.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const FollowWaypoints_ResultRefStruct = StructType({
  status: GoalStatusWrapper.refObjectType,
  missed_waypoints: Int32Wrapper.refObjectArrayType,
});
const FollowWaypoints_ResultRefArray = ArrayType(FollowWaypoints_ResultRefStruct);
const FollowWaypoints_ResultRefStructArray = StructType({
  data: FollowWaypoints_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowWaypoints_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowWaypoints_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.status =  new GoalStatusWrapper(other._wrapperFields.status);
      this._wrapperFields.missed_waypoints = Int32Wrapper.createArray();
      this._wrapperFields.missed_waypoints.copy(other._wrapperFields.missed_waypoints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowWaypoints_ResultRefStruct();
    this._wrapperFields.status = new GoalStatusWrapper();
    this._wrapperFields.missed_waypoints = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new FollowWaypoints_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowWaypoints_ResultArrayWrapper;
  }
  static get ArrayType() {
    return FollowWaypoints_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowWaypoints_ResultRefStructArray
  }
  static get refObjectType() {
    return FollowWaypoints_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    this._wrapperFields.missed_waypoints.freeze(own, checkConsistency);
    this._refObject.missed_waypoints = this._wrapperFields.missed_waypoints.refObject;
    if (own) {
      this._wrapperFields.missed_waypoints.fill(Int32Array.from([]));
      this._wrapperFields.missed_waypoints.freeze(own, checkConsistency);
      this._refObject.missed_waypoints = this._wrapperFields.missed_waypoints.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._wrapperFields.missed_waypoints.copyRefObject(refObject.missed_waypoints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GoalStatusWrapper.freeStruct(refObject.status);
    if (refObject.missed_waypoints.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.missed_waypoints);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        deallocator.delayFreeStructMember(refObject.missed_waypoints, Int32Wrapper.refObjectArrayType, 'data');
      } else {
        deallocator.freeStructMember(refObject.missed_waypoints, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    FollowWaypoints_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'FollowWaypoints_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof GoalStatusWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new GoalStatusWrapper(value));
    }
  }
  get missed_waypoints() {
    return this._wrapperFields['missed_waypoints'].data;
  }
  set missed_waypoints(value) {
    this._wrapperFields['missed_waypoints'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new FollowWaypoints_ResultRefStruct(refObject.toObject());
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._wrapperFields.missed_waypoints.copyRefObject(this._refObject.missed_waypoints);
  }
  copy(other) {
    this._refObject = new FollowWaypoints_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._wrapperFields.missed_waypoints.copy(other._wrapperFields.missed_waypoints);
  }
  static get classType() {
    return FollowWaypoints_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"action_msgs","type":"GoalStatus","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"missed_waypoints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"FollowWaypoints_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowWaypoints_Result"};
  }
  hasMember(name) {
    let memberNames = ["status","missed_waypoints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowWaypoints_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowWaypoints_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowWaypoints_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowWaypoints_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowWaypoints_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowWaypoints_ResultRefArray(size);
    this._refObject = new FollowWaypoints_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowWaypoints_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowWaypoints_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowWaypoints_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowWaypoints_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowWaypoints_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowWaypoints_ResultArrayWrapper;
  }
}
module.exports = FollowWaypoints_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "action_msgs",
        "type": "GoalStatus",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "missed_waypoints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "FollowWaypoints_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowWaypoints_Result"
}
*/
