// rclnodejs/generated/nav2_msgs/nav2_msgs__action__NavigateToPose.js
'use strict';
class nav2_msgs__action__NavigateToPose {
  static get Goal() {
    return require('./nav2_msgs__action__NavigateToPose_Goal.js');
  }
  static get GoalRes() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get Cancel() {
    return require('../action_msgs/action_msgs__srv__CancelGoalRequest.js');
  }
  static get CancelRes() {
    return require('../action_msgs/action_msgs__srv__CancelGoalResponse.js');
  }
  static get ResultReq() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get ResultRes() {
    return require('./nav2_msgs__action__NavigateToPose_Result.js');
  }
  static get Feedback() {
    return require('./nav2_msgs__action__NavigateToPose_Feedback.js');
  }
  static get Status() {
    return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'NavigateToPose'};
  }
}
module.exports = nav2_msgs__action__NavigateToPose;
