// rclnodejs/generated/tf2_msgs/tf2_msgs__action__LookupTransform.js
'use strict';
class tf2_msgs__action__LookupTransform {
  static get Goal() {
    return require('./tf2_msgs__action__LookupTransform_Goal.js');
  }
  static get GoalRes() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get Cancel() {
    return require('../action_msgs/action_msgs__srv__CancelGoalRequest.js');
  }
  static get CancelRes() {
    return require('../action_msgs/action_msgs__srv__CancelGoalResponse.js');
  }
  static get ResultReq() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get ResultRes() {
    return require('./tf2_msgs__action__LookupTransform_Result.js');
  }
  static get Feedback() {
    return require('./tf2_msgs__action__LookupTransform_Feedback.js');
  }
  static get Status() {
    return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
  }
  static type() {
    return {pkgName: 'tf2_msgs', subFolder: 'action', interfaceName: 'LookupTransform'};
  }
}
module.exports = tf2_msgs__action__LookupTransform;
