// rclnodejs/generated/test_msgs/test_msgs__action__NestedMessage.js
'use strict';
class test_msgs__action__NestedMessage {
  static get Goal() {
    return require('./test_msgs__action__NestedMessage_Goal.js');
  }
  static get GoalRes() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get Cancel() {
    return require('../action_msgs/action_msgs__srv__CancelGoalRequest.js');
  }
  static get CancelRes() {
    return require('../action_msgs/action_msgs__srv__CancelGoalResponse.js');
  }
  static get ResultReq() {
    return require('../action_msgs/action_msgs__msg__GoalInfo.js');
  }
  static get ResultRes() {
    return require('./test_msgs__action__NestedMessage_Result.js');
  }
  static get Feedback() {
    return require('./test_msgs__action__NestedMessage_Feedback.js');
  }
  static get Status() {
    return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'action', interfaceName: 'NestedMessage'};
  }
}
module.exports = test_msgs__action__NestedMessage;
